src/dd/dd.rs

```diff
@@ -0,0 +1,679 @@
+        }
+    }
+    inp
+}
+
+fn setup_output(cf: Arc<Mutex<Conf>>) -> OutFileWrap {
+    let cf_guard = cf.lock().unwrap();
+
+    let out = if cf_guard.is_stdout {
+        Some(OutFileWrap {
+            f: OutFileType::Stdout(io::stdout()), // TODO: use .lock() if necessary
+            is_seekable: false,
+        })
+    } else {
+        match File::create(&cf_guard.out_name) {
+            Err(e) => {
+                err(format!("cannot open output file: {}", e.description()));
+                None
+            }
+            Ok(f) => {
+                Some(OutFileWrap {
+                    f: OutFileType::Reg(f),
+                    is_seekable: true, // TODO: handle case when device isn't seekable
+                })
+            }
+        }
+    };
+    let mut out = out.unwrap();
+
+    let mut opts = OpenOptions::new();
+    opts.read(false).write(true);
+
+    let zero_buf: Vec<u8> = vec![0; 4096];
+    match out.f {
+        OutFileType::Reg(ref mut f) => {
+            if out.is_seekable {
+                if cf_guard.seek == 0 {
+                    if !cf_guard.has_flag(Conf::NOTRUNC) {
+                        if f.seek(io::SeekFrom::Start(cf_guard.seek as u64)).is_err() {
+                            err("cannot seek on output file".to_string());
+                        }
+                    }
+                }
+            } else {
+                let mut to_skip = cf_guard.seek * cf_guard.out_block_size;
+                let mut buf = vec![0u8; BUF_SIZE];
+                while to_skip > 0 {
+                    match f.read(&mut buf[..cmp::min(to_skip, BUF_SIZE)]) {
+                        Ok(0) => break,
+                        Ok(m) => to_skip -= m,
+                        _ => err("cannot seek on output file".to_string()),
+                    }
+                }
+                if to_skip > 0 {
+                    while to_skip > 0 {
+                        match f.write(&zero_buf[..cmp::min(to_skip, BUF_SIZE)]) {
+                            Ok(0) => err("cannot seek on output file".to_string()),
+                            Ok(m) => to_skip -= m,
+                            _ => err("cannot seek on output file".to_string()),
+                        }
+                    }
+                }
+            }
+        }
+
+        // Would have terminated early if seek != 0, test precondition anyway
+        OutFileType::Stdout(_) => assert_eq!(cf_guard.seek, 0),
+    }
+    out
+}
+
+// , _sdone: chan::Sender<()>
+fn setup(cf: Arc<Mutex<Conf>>) -> (InpFileWrap, OutFileWrap) {
+    let inp = setup_input(Arc::clone(&cf));
+    let out = setup_output(cf);
+
+    (inp, out)
+}
+
+fn copy(
+    mut inp: InpFileWrap,
+    mut out: OutFileWrap,
+    cf_ptr: Arc<Mutex<Conf>>,
+    _sdone: chan::Sender<()>,
+) {
+    {
+        let mut cf = cf_ptr.lock().unwrap();
+        let mut buf = vec![0u8; cmp::min(cf.inp_block_size, cf.out_block_size)];
+        let mut total = 0;
+        for _ in 0..cf.count {
```

Comment:
This should behave like an infinite loop which breaks when everything has been copied if `cf.count == None`.

Reply:
Also to avoid the display status thread from blocking until the end, you are definitely going to need to reacquire the mutex sometimes (either every loop iteration or every few seconds or something like that).  You could also consider using a `RwLock` as the main thread is the only one that modifies the struct (and it only gets modified in a couple spots in this function).

Reply:
Changed.